# Bike Sharing Demand Analysis: Understanding Rental Patterns to Improve Operational Efficiency


## Define Business Questions


- **Question 1: How has bike rental demand changed over time?**
- **Question 2: Which season generates the highest number of rentals?**
- **Question 3: How do different weather conditions affect bike rental demand?**
- **Question 4: At what hours of the day is bike demand highest?**
- **Question 5: How does rental demand differ between working days, weekends, and holidays?**
- **Question 6: Which variables have the strongest relationship with bike rentals?**


## Import Packages/Library 

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from datetime import time

warnings.filterwarnings("ignore")

## Data Wrangling


### Gathering Data


#### Load daily_df

In [2]:
# Load daily data
daily_df = pd.read_csv("./data/day.csv")
daily_df.head()

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,6,0,2,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,2011-01-02,1,0,1,0,0,0,2,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,2011-01-03,1,0,1,0,1,1,1,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,2011-01-04,1,0,1,0,2,1,1,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,2011-01-05,1,0,1,0,3,1,1,0.226957,0.229270,0.436957,0.186900,82,1518,1600


The dataset contains various columns, including:

- **instant**: Record index.
- **dteday**: Date of the record.
- **season**: The season in which the data was recorded (1: Spring, 2: Summer, 3: Fall, 4: Winter).
- **yr**: Year of the data (0: 2011, 1: 2012).
- **mnth**: Month of the year (1 to 12).
- **holiday**: Whether the day is a holiday (1: Yes, 0: No).
- **weekday**: Day of the week (0: Sunday, 6: Saturday).
- **workingday**: Whether the day is a working day (1: Yes, 0: No).
- **weathersit**: Weather situation (1: Clear, 2: Mist, 3: Light rain/snow, 4: Heavy rain/snow).
- **temp** : Normalized temperature in Celsius. The values are divided to 41 (max).
- **atemp**: Normalized feeling temperature in Celsius. The values are divided to 50 (max).
- **hum**: Normalized humidity. The values are divided to 100 (max).
- **windspeed**: Normalized wind speed. The values are divided to 67 (max).
- **casual**: Count of casual users.
- **registered**: Count of registered users.
- **cnt**: Count of total rental bikes including both casual and registered.

By examining the first few rows, we can get an overview of the dataset before performing further analysis.


#### Load hourly_df

In [3]:
# Load hour data
hourly_df = pd.read_csv("./data/hour.csv")
hourly_df.head()

,instant,dteday,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,0,6,0,1,0.24,0.2879,0.81,0.0,3,13,16
1,2,2011-01-01,1,0,1,1,0,6,0,1,0.22,0.2727,0.80,0.0,8,32,40
2,3,2011-01-01,1,0,1,2,0,6,0,1,0.22,0.2727,0.80,0.0,5,27,32
3,4,2011-01-01,1,0,1,3,0,6,0,1,0.24,0.2879,0.75,0.0,3,10,13
4,5,2011-01-01,1,0,1,4,0,6,0,1,0.24,0.2879,0.75,0.0,0,1,1


This dataset contains similar features to the daily dataset (`day_df`), but with hourly records instead of daily aggregates. Some key columns include:

- **instant**: Unique identifier for each row.
- **dteday**: Date of the record.
- **hr**: Hour of the day (0 to 23).
- **season, yr, mnth, holiday, weekday, workingday, weathersit**: Similar to the daily dataset, representing seasonal and time-related information.
- **temp, atemp, hum, windspeed**: Weather-related features.
- **casual, registered, cnt**: Number of casual users, registered users, and total bike rentals for each hour.

Since this dataset provides a more granular view with hourly records, it allows for a deeper analysis of bike-sharing trends throughout the day.


### Assessing Data


#### Identifying daily_df problems

In [4]:
# Day data information
daily_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    str    
 2   season      731 non-null    int64  
 3   yr          731 non-null    int64  
 4   mnth        731 non-null    int64  
 5   holiday     731 non-null    int64  
 6   weekday     731 non-null    int64  
 7   workingday  731 non-null    int64  
 8   weathersit  731 non-null    int64  
 9   temp        731 non-null    float64
 10  atemp       731 non-null    float64
 11  hum         731 non-null    float64
 12  windspeed   731 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(4), int64(11), str(1)
memory usage: 98.6 KB


- The dataset contains **731 rows** and **16 columns**.
- All columns have **731 non-null values**, meaning there are **no missing values**.
- The dataset consists of different data types:
  - **Integer (`int64`)**: Used for categorical and count-based features such as `season`, `yr`, `mnth`, `holiday`, `weekday`, `workingday`, `weathersit`, `casual`, `registered`, and `cnt`.
  - **Float (`float64`)**: Used for continuous numerical values like `temp`, `atemp`, `hum`, and `windspeed`.
  - **Object (`object`)**: The `dteday` column, which represents dates in string format.

This information helps in preprocessing and deciding if any columns need transformation (e.g., converting `dteday` to a proper datetime format).


In [5]:
# Check Duplicate the day data
daily_df.duplicated().sum()

np.int64(0)

The output is **0**, which means there are no duplicate records in the dataset. This confirms that each row represents a unique entry, and no additional cleaning is required for duplicates.


In [6]:
# Check Missing Value the day data
daily_df.isna().sum()

instant       0
dteday        0
season        0
yr            0
mnth          0
holiday       0
weekday       0
workingday    0
weathersit    0
temp          0
atemp         0
hum           0
windspeed     0
casual        0
registered    0
cnt           0
dtype: int64

The output shows that **all columns have 0 missing values**, meaning the dataset is complete and does not require imputation or removal of missing data.

Since there are no missing values, we can proceed with the analysis without additional preprocessing for missing data.


In [ ]:
# The descriptive statistic from day data
daily_df.describe()

To understand the overall distribution of numerical variables, we use the `.describe()` function. This provides key statistical summaries, including:

- **count**: Number of non-null values in each column.
- **mean**: Average value of each column.
- **std (standard deviation)**: Measures data dispersion from the mean.
- **min and max**: Minimum and maximum values.
- **25%, 50% (median), and 75%**: Quartiles, showing data distribution.

Key Observations:

- The dataset consists of **731 records**.
- The **average number of total bike rentals (`cnt`) is 4504**, with a **minimum of 22** and a **maximum of 8714**.
- The **temperature (`temp`) ranges from 0.06 to 0.86** (normalized values).
- The dataset is **balanced across years (`yr`)**, with values of `0` (2011) and `1` (2012).
- The **working day (`workingday`) and holiday (`holiday`) columns are binary**, indicating whether a day is a working day or holiday.

This summary helps in understanding the overall trends and potential data distributions before further analysis.


#### Identifying hour_df problems

In [ ]:
# Hour data information
hourly_df.info()

To understand the structure of the hourly dataset, we use the `.info()` function. This provides an overview of the dataset, including:

- **Total entries**: 17,379 rows.
- **Total columns**: 17 attributes.
- **Data types**:
  - `int64`: 12 columns (e.g., `season`, `yr`, `mnth`, `hr`, `holiday`, etc.).
  - `float64`: 4 columns (`temp`, `atemp`, `hum`, `windspeed`).
  - `object`: 1 column (`dteday`, representing dates).
- **Memory usage**: 2.3+ MB.

Key Observations:

- The dataset contains **hourly records of bike rentals** over two years.
- There are **no missing values** in any column, which simplifies data preprocessing.
- The `hr` column represents hours of the day (0-23), allowing for time-based analysis.
- `cnt` is the total number of rentals, obtained from the sum of `casual` and `registered` users.

This summary helps in understanding the dataset before performing further exploratory data analysis.


In [ ]:
# Check duplicate the hour data
hourly_df.duplicated().sum()

The output is **0**, meaning there are **no duplicate entries** in the dataset. This confirms that each row represents a unique hourly record without redundancy.


In [ ]:
# Check Missing Value the hour data
hourly_df.isnull().sum()

The output shows **0 missing values** in all 17 columns. This means the dataset is **fully complete**, with no need for imputation or data cleaning related to missing values.

In [ ]:
# The descriptive statistic from day data
hourly_df.describe().round(2)

To understand the distribution of numerical features, we use the `.describe().round(2)` function. This provides key statistical summaries, including count, mean, standard deviation, min, max, and quartiles.

Key Observations:

- **Total Entries**: 17,379 rows.
- **Hourly Distribution** (`hr`): The values range from **0 to 23**, covering all hours in a day.
- **Temperature (`temp`) & Feels-Like Temperature (`atemp`)**:
  - `temp` (Normalized between 0 and 1) has a **mean of 0.50**, suggesting an even spread over different weather conditions.
  - `atemp` follows a similar distribution with a mean of **0.48**.
- **Humidity (`hum`) & Wind Speed (`windspeed`)**:
  - Humidity ranges from **0 to 1**, with a mean of **0.63**.
  - Wind speed has an average of **0.19**, indicating most days experience mild winds.
- **Bike Rentals (`cnt`)**:
  - The total count (`cnt`) ranges from **1 to 977** per hour.
  - Median rentals per hour is **142**, while the upper quartile reaches **281**, indicating some peak hours with high demand.
  - `casual` riders (mean **35.68**) are significantly fewer than `registered` riders (mean **153.79**), suggesting that most users are registered members.

Conclusion:  
This statistical summary provides insight into the distribution and variability of bike rentals across different hours, weather conditions, and user types. These insights will guide further analysis and model building.


### Cleaning Data


#### Fixing hourly_data problem

In [ ]:
hourly_df.head()

In [ ]:
# Change the column names
hourly_column = {
    "instant": "index",
    "dteday": "date",
    "yr": "year",
    "mnth": "month",
    "hr": "hour",
    "temp": "temperature",
    "atemp": "feels_like_temperature",
    "hum": "humidity",
    "cnt": "count",
}

hourly_df.rename(columns=hourly_column, inplace=True)

In [ ]:
# Fix the datetime data
hourly_df["date"] = pd.to_datetime(hourly_df["date"])

In [ ]:
# Categorical season column
season = {1: "springer", 2: "summer", 3: "fall", 4: "winter"}

hourly_df["season"] = hourly_df["season"].replace(season)

In [ ]:
# Year column
year = {0: 2011, 1: 2012}

hourly_df["year"] = hourly_df["year"].replace(year)

In [ ]:
# Month column
month = {
    1: "January",
    2: "February",
    3: "March",
    4: "April",
    5: "May",
    6: "June",
    7: "July",
    8: "August",
    9: "September",
    10: "October",
    11: "November",
    12: "December",
}

hourly_df["month"] = hourly_df["month"].replace(month)

In [ ]:
# Hour Column
hourly_df["hour"] = hourly_df["hour"].apply(lambda x: time(x, 0))

In [ ]:
# Change holiday column to categorical values
holiday = {0: "No", 1: "Yes"}

hourly_df["holiday"] = hourly_df["holiday"].replace(holiday)

In [ ]:
# Change coloumns weekdays from numerical to categorical
weekdays = {
    0: "Sunday",
    1: "Monday",
    2: "Tuesday",
    3: "Wednesday",
    4: "Thursday",
    5: "Friday",
    6: "Saturday",
}

hourly_df["weekday"] = hourly_df["weekday"].replace(weekdays)

In [ ]:
# Change coloumns workingday from numerical to categorical
workingday = {0: "No", 1: "Yes"}

hourly_df["workingday"] = hourly_df["workingday"].replace(workingday)

In [ ]:
# Change the weathersit from numerical to categorical
weathersit_labels = {
    1: "Clear",
    2: "Cloudy or Misty",
    3: "Light Rain/Snow",
    4: "Heavy Rain/Snow",
}

hourly_df["weathersit"] = hourly_df["weathersit"].map(weathersit_labels)

In [ ]:
# Denormalized temperature column with multiple by 41
hourly_df["temperature"] = hourly_df["temperature"] * 41

In [ ]:
# Denormalized feels_like_temperature column with multiple by 50
hourly_df["feels_like_temperature"] = hourly_df["feels_like_temperature"] * 50

In [ ]:
# Denormalized humidity column with multiple by 100
hourly_df["humidity"] = hourly_df["humidity"] * 100

In [ ]:
# Denormalized windspeed column with multiple by 67
hourly_df["windspeed"] = hourly_df["windspeed"] * 67

In [ ]:
hourly_df.head()

#### Fixing daily_df problem

In [ ]:
daily_df.head()

In [ ]:
# Change the column names
daily_columns = {
    "instant": "index",
    "dteday": "date",
    "yr": "year",
    "mnth": "month",
    "temp": "temperature",
    "atemp": "feels_like_temperature",
    "hum": "humidity",
    "cnt": "count",
}

daily_df.rename(columns=daily_columns, inplace=True)

In [ ]:
# Fix the datetime data
daily_df["date"] = pd.to_datetime(daily_df["date"])

In [ ]:
# Categorical season column
season = {1: "springer", 2: "summer", 3: "fall", 4: "winter"}

daily_df["season"] = daily_df["season"].replace(season)

In [ ]:
# Year column
year = {0: 2011, 1: 2012}

daily_df["year"] = daily_df["year"].replace(year)

In [ ]:
# Month column
month = {
    1: "January",
    2: "February",
    3: "March",
    4: "April",
    5: "May",
    6: "June",
    7: "July",
    8: "August",
    9: "September",
    10: "October",
    11: "November",
    12: "December",
}

daily_df["month"] = daily_df["month"].replace(month)

In [ ]:
# Change holiday column to categorical values
holiday = {0: "No", 1: "Yes"}

daily_df["holiday"] = daily_df["holiday"].replace(holiday)

In [ ]:
# Change coloumns weekdays from numerical to categorical
weekdays = {
    0: "Sunday",
    1: "Monday",
    2: "Tuesday",
    3: "Wednesday",
    4: "Thursday",
    5: "Friday",
    6: "Saturday",
}

daily_df["weekday"] = daily_df["weekday"].replace(weekdays)

In [ ]:
# Change coloumns workingday from numerical to categorical
workingday = {0: "No", 1: "Yes"}

daily_df["workingday"] = daily_df["workingday"].replace(workingday)

In [ ]:
# Change the weathersit from numerical to categorical
weathersit_labels = {
    1: "Clear",
    2: "Cloudy or Misty",
    3: "Light Rain/Snow",
    4: "Heavy Rain/Snow",
}

daily_df["weathersit"] = daily_df["weathersit"].map(weathersit_labels)

In [ ]:
# Denormalized temperature column with multiple by 41
daily_df["temperature"] = daily_df["temperature"] * 41

In [ ]:
# Denormalized feels_like_temperature column with multiple by 50
daily_df["feels_like_temperature"] = daily_df["feels_like_temperature"] * 50

In [ ]:
# Denormalized windspeed column with multiple by 67
daily_df["windspeed"] = daily_df["windspeed"] * 67

In [ ]:
daily_df.head()

## Exploratory Data Analysis (EDA)


### Explore daily_df


**Insight:** (Opsional)

- xxx
- xxx


## Visualization & Explanatory Analysis


### Pertanyaan 1:


### Pertanyaan 2:


**Insight:** (Opsional)

- xxx
- xxx


## Analisis Lanjutan (Opsional)


**Insight:** (Opsional)

- xxx
- xxx


## Conclusion & Recommendation


- **Conclusion pertanyaan 1:** xxx
- **Conclusion pertanyaan 2:** xxx
- ...


**Rekomendasi Action Item:**

- xxx
- xxx
